src/Token.php

```diff
@@ -9,267 +9,59 @@
-     */
-    public function __construct(
-        array $headers = ['alg' => 'none'],
-        array $claims = [],
-        Signature $signature = null,
-        array $payload = ['', '']
-    ) {
-        $this->headers = $headers;
-        $this->claims = $claims;
-        $this->signature = $signature;
-        $this->payload = $payload;
-    }
-
-    /**
      * Returns the token headers
-     *
-     * @return array
      */
-    public function getHeaders(): array
-    {
-        return $this->headers;
-    }
+    public function headers(): DataSet;
 
     /**
-     * Returns if the header is configured
-     *
-     * @param string $name
-     *
-     * @return bool
+     * Returns if the token is allowed to be used by the audience
      */
-    public function hasHeader(string $name): bool
-    {
-        return array_key_exists($name, $this->headers);
-    }
+    public function isAllowedTo(string $audience): bool;
 
     /**
-     * Returns the value of a token header
-     *
-     * @param string $name
-     * @param mixed $default
-     *
-     * @return mixed
-     *
-     * @throws OutOfBoundsException
+     * Returns if the token has the given id
      */
-    public function getHeader(string $name, $default = null)
-    {
-        if ($this->hasHeader($name)) {
-            return $this->getHeaderValue($name);
-        }
-
-        if ($default === null) {
-            throw new OutOfBoundsException('Requested header is not configured');
-        }
-
-        return $default;
-    }
+    public function isIdentifiedBy(string $id): bool;
 
     /**
-     * Returns the value stored in header
-     *
-     * @param string $name
-     *
-     * @return mixed
+     * Returns if the token has the given subject
      */
-    private function getHeaderValue(string $name)
-    {
-        $header = $this->headers[$name];
-
-        if ($header instanceof Claim) {
-            return $header->getValue();
-        }
-
-        return $header;
-    }
+    public function isRelatedTo(string $subject): bool;
 
     /**
-     * Returns the token claim set
-     *
-     * @return array
+     * Returns if the token was issued by any of given issuers
      */
-    public function getClaims(): array
-    {
-        return $this->claims;
-    }
+    public function hasBeenIssuedBy(string ...$issuers): bool;
 
     /**
-     * Returns if the claim is configured
-     *
-     * @param string $name
-     *
-     * @return bool
+     * Returns if the token was issued before of given time
      */
-    public function hasClaim(string $name): bool
-    {
-        return array_key_exists($name, $this->claims);
-    }
+    public function hasBeenIssuedBefore(DateTimeInterface $now): bool;
 
     /**
-     * Returns the value of a token claim
-     *
-     * @param string $name
-     * @param mixed $default
-     *
-     * @return mixed
-     *
-     * @throws OutOfBoundsException
+     * Returns if the token minimum time is before than given time
```

Comment:
Ah, that's confusing because of the `not` in `not before` :-\

I guess that's what the spec is - can't fix the naming then.